flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp

```diff
@@ -154,6 +156,56 @@ mlir::LLVM::DITypeAttr DebugTypeGenerator::convertBoxedSequenceType(
       dataLocation, /*rank=*/nullptr, allocated, associated);
 }
 
+mlir::LLVM::DITypeAttr DebugTypeGenerator::convertRecordType(
```

Comment:
This processing is likely not cheap, do you know how many times this will being called for a given derived type in a program: once per type, or once for every variable with this type, or more?

It may be worth having some caching `fir::RecordType Ty` <-> `mlir::LLVM::DITypeAttr`, or even `mlir::Type` <-> `mlir::LLVM::DITypeAttr`

Reply:
Keeping a cache for `mlir::Type` to `mlir::DITypeAttr` was always the plan. Although I have noticed in the IR dump that there is no type duplication. So may be mlir does caching at some level. I will keep this thing as a TODO to avoid generating redundant types.